File: rudder-sdk-core/src/main/java/com/rudderlabs/android/sdk/core/RudderProperty.java
package com.rudderlabs.android.sdk.core;

import com.google.gson.Gson;
import com.rudderlabs.android.sdk.core.util.Utils;

import java.util.HashMap;
import java.util.Map;

public class RudderProperty {
    private Map<String, Object> map = new HashMap<>();

    Map<String, Object> getMap() {
        return map;
    }

    public boolean hasProperty(String key) {
        return map.containsKey(key);
    }

    public Object getProperty(String key) {
        return map.containsKey(key) ? map.get(key) : null;
    }

    public void put(String key, Object value) {
        map.put(key, value);
    }

    public RudderProperty putValue(String key, Object value) {
        map.put(key, value);
        return this;
    }

    public RudderProperty putValue(Map<String, Object> map) {
        if (map != null) this.map.putAll(map);
        return this;
    }

    public void putRevenue(double revenue) {
        map.put("revenue", revenue);
    }

    public void putCurrency(String currency) {
        map.put("currency", currency);
    }

    public void putCurrency(String currency) {
        map.put("currency", currency);
    }
}

File: models/src/main/java/com/rudderstack/models/android/RudderProperty.kt
package com.rudderstack.models.android

import com.rudderstack.models.android.RudderProperty
import java.util.HashMap

abstract class RudderProperty {
    private val map: MutableMap<String, Any> = HashMap()
    open fun getMap(): Map<String, Any> {
        return map
    }

    fun hasProperty(key: String): Boolean {
        return map.containsKey(key)
    }

    fun getProperty(key: String): Any? {
        return if (map.containsKey(key)) map[key] else null
    }

    fun put(key: String, value: Any) {
        map[key] = value
    }

    fun putValue(key: String, value: Any): RudderProperty {
        if (value is RudderProperty) {
            map[key] = value.getMap()
        } else {
            map[key] = value
        }
        return this
    }

    fun putValue(map: Map<String, Any>?): RudderProperty {
        if (map != null) this.map.putAll(map)
        return this
    }

    fun putRevenue(revenue: Double) {
        map["revenue"] = revenue
    }

    fun putCurrency(currency: String) {
        map["currency"] = currency
    }
}
data class ScreenProperty(private val screenName : String, private val isAutomatic : Boolean = false) :
    RudderProperty() {
    init {
        putValue("name", screenName)
        putValue("automatic", isAutomatic)
    }
}